ShareIt v2.0 add-controllers

Zazergel opened the pull request:
Реализовал базу приложения

Zazergel committed:
ShareIt v1.0 add-controllers
Реализовал базу приложения

Zazergel committed:
ShareIt v1.1 add-controllers
fix: code-style #1

NikolaiDem commented on src/main/java/ru/practicum/shareit/exceptions/handler/ErrorHandler.java:
```diff
@@ -0,0 +1,81 @@
+    ) {
+        final List<Violation> violations = e.getConstraintViolations().stream()
+                .map(
+                        violation -> new Violation(
```
Перенос на следующую строку сделан преднамеренно ?)

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/controller/ItemController.java:
```diff
@@ -0,0 +1,52 @@
+package ru.practicum.shareit.item.controller;
```
Так как в пакете item предполагается по одному контроллеру, сервису, модели и т.д., то создавать под каждый класс подпакет выглядит избыточно.
Аналогично для остальных пакетов, в которых содержится по одному классу.

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/controller/ItemController.java:
```diff
@@ -0,0 +1,52 @@
+ * TODO Sprint add-controllers.
+ */
+@AllArgsConstructor
+@Validated
```
Аннотация необходима, если в классе определены constraint annotations, например из javax.validation.constraints

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/controller/ItemController.java:
```diff
@@ -0,0 +1,52 @@
+        if (userId != null) {
+            return itemService.getByOwner(userId);
+        } else {
+            return itemService.getAll();
```
По спецификации эндпоинт должен вернуть только список вещей владельца

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/dao/ItemStorage.java:
```diff
@@ -0,0 +1,17 @@
+
+@Component
+@AllArgsConstructor
+@Data
```
Для хранилища эти методы лишние.

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/dto/ItemMapper.java:
```diff
@@ -0,0 +1,19 @@
+import org.springframework.stereotype.Component;
+import ru.practicum.shareit.item.model.Item;
+
+@Component
```
Делать бином избыточно, методы класса статические.

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -1,7 +1,37 @@
+@Builder
 public class Item {
+    private Long id;
+    @NotBlank(message = "Название предмета не должно быть пустым.")
```
В модели валидации нет, constraint annotations лишние, аналогично для остальных полей.

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -1,7 +1,37 @@
+    private Long request;
+
+    @Override
+    public boolean equals(Object o) {
```
Для чего переопределяется equals и hashCode?

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/service/ItemService.java:
```diff
@@ -0,0 +1,21 @@
+
+import java.util.List;
+
+public interface ItemService {
```
В спецификации методов сервиса лучше использовать dto объекты, а не объекты модели. Модель - это внутреннее представление БД, с которой работает репозиторий. А dto - это объект для общения с внешним миром.

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,148 @@
+        ownerCheck(owner);
+        item.setOwner(owner);
+        if (item.getId() != null) {
+            if (itemStorage.getItemStorage().containsKey(item.getId())) {
```
Обращаться ко внутреннему представлению хранилища не нужно. У него должны быть определены методы existsById(), save(), update(), get() и т.д., которые может вызывать сервис. Делается для того, чтобы не зависеть от реализации хранилища.
Аналогично для остальных методов.

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,148 @@
+            Item item = itemStorage.getItemStorage().get(itemId);
+            log.info("Предмет '" + item.getName() + "' id=" + item.getId() + " был(-а) успешно получен(-а).");
+            return item;
+        } catch (NullPointerException e) {
```
Перехватывать runtime исключения в бизнес коде нехорошо.
Исправить везде.

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,148 @@
+                throw new AuthorizationException(message);
+            }
+            if (item.getName() != null) {
+                if (!item.getName().isBlank()) {
```
Для чего вложенные проверки?

NikolaiDem commented on src/main/java/ru/practicum/shareit/user/controller/UserController.java:
```diff
@@ -0,0 +1,46 @@
+package ru.practicum.shareit.user.controller;
```
Так как в пакете user предполагается по одному контроллеру, сервису, модели и т.д., то создавать под каждый класс подпакет выглядит избыточно.
Аналогично для остальных пакетов, в которых содержится по одному классу.

NikolaiDem commented on src/main/java/ru/practicum/shareit/user/controller/UserController.java:
```diff
@@ -0,0 +1,46 @@
+
+
+@AllArgsConstructor
+@Validated
```
Аннотация необходима, если в классе определены constraint annotations, например из javax.validation.constraints

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/dao/ItemStorage.java:
```diff
@@ -0,0 +1,17 @@
+@AllArgsConstructor
+@Data
+public class ItemStorage {
+    private final HashMap<Long, Set<Long>> userItemsId;
```
Map<Long, Set<Long>>

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/dao/ItemStorage.java:
```diff
@@ -0,0 +1,17 @@
+@Data
+public class ItemStorage {
+    private final HashMap<Long, Set<Long>> userItemsId;
+    private final HashMap<Long, Item> itemStorage;
```
Map<Long, Item>

NikolaiDem commented on src/main/java/ru/practicum/shareit/user/dao/UserStorage.java:
```diff
@@ -0,0 +1,18 @@
+@Data
+public class UserStorage {
+
+    private final HashMap<Long, User> userStorage;
```
Map<Long, User>

NikolaiDem commented on src/main/java/ru/practicum/shareit/user/dto/UserDto.java:
```diff
@@ -0,0 +1,14 @@
+
+    private Long id;
+    private String name;
+}
```
А email?

NikolaiDem commented on src/main/java/ru/practicum/shareit/user/dto/UserMapper.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.stereotype.Component;
+import ru.practicum.shareit.user.model.User;
+
+@Component
```
Делать бином избыточно, методы класса статические.

NikolaiDem commented on src/main/java/ru/practicum/shareit/user/dto/UserMapper.java:
```diff
@@ -0,0 +1,14 @@
+        return UserDto.builder()
+                .id(user.getId())
+                .name(user.getName())
+                .build();
```
А email?

NikolaiDem commented on src/main/java/ru/practicum/shareit/user/model/User.java:
```diff
@@ -0,0 +1,24 @@
+public class User {
+
+    private Long id;
+    @NotBlank(message = "Имя пользователя не должно быть пустым.")
```
В модели валидации нет, constraint annotations лишние, аналогично для остальных полей.

NikolaiDem commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -1,7 +1,37 @@
+import javax.validation.constraints.NotNull;
+import java.util.Objects;
+
+@Data
```
Для чего Data?

NikolaiDem commented on src/main/java/ru/practicum/shareit/user/model/User.java:
```diff
@@ -0,0 +1,24 @@
+/**
+ * TODO Sprint add-controllers.
+ */
+@Data
```
Для чего Data?

NikolaiDem commented on src/main/java/ru/practicum/shareit/user/service/UserService.java:
```diff
@@ -0,0 +1,18 @@
+
+import java.util.List;
+
+public interface UserService {
```
В спецификации методов сервиса лучше использовать dto объекты, а не объекты модели. Модель - это внутреннее представление БД, с которой работает репозиторий. А dto - это объект для общения с внешним миром.

NikolaiDem commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,107 @@
+    public User add(User user) {
+        emailCheck(user);
+        if (user.getId() != null) {
+            if (userStorage.getUserStorage().containsKey(user.getId())) {
```
Обращаться ко внутреннему представлению хранилища не нужно. У него должны быть определены методы existsById(), save(), update(), get() и т.д., которые может вызывать сервис. Делается для того, чтобы не зависеть от реализации хранилища.
Аналогично для остальных методов.

NikolaiDem commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,107 @@
+            User user = userStorage.getUserStorage().get(userId);
+            log.info("Пользователь id=" + userId + " успешно получен.");
+            return user;
+        } catch (NullPointerException e) {
```
Перехватывать runtime исключения в бизнес коде нехорошо.
Исправить везде.

Zazergel committed:
ShareIt v2.0 add-controllers
feat: Полностью переработана логика приложения (почти что с нуля), теперь операции с моделями берут на себя хранилища (валидация, создание, хранение, изменение, поиск, присвоение id и т.д.)
fix: Переработаны аннотации в классах - моделях.
feat: Дополнены "мапперы" моделей.
fix: Теперь контроллеры оперируют Dto объектами, а не моделями.
fix: Немного переработан ErrorHandler
fix: исправлена опечатка поля "broker" в классе Booking.java на "bookerId"
fix: code-style #2
refactor: пакеты, содержащие всего один класс были удалены, классы извлечены в пакет повыше.
delete: удалены лишние методы equals & hashcode в классах - моделях
delete: удалены лишние аннотации @Data.
